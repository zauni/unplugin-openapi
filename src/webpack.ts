import { unplugin } from "./index.js";

export default unplugin.webpack as typeof unplugin.webpack;
