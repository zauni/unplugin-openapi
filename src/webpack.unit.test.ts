import { Volume, createFsFromVolume } from "memfs";
import { describe, expect, it } from "vitest";
import webpack, { type OutputFileSystem } from "webpack";
import webpackPlugin from "./webpack.js";

async function build(config: webpack.Configuration): Promise<string> {
	return new Promise((resolve, reject) => {
		const compiler = webpack({
			mode: "production",
			output: {
				path: "/", // Set to '/' since we are using an in-memory filesystem
				filename: "bundle.js", // Output file name
			},
			...config,
		});
		const vol = new Volume();
		const memFs = createFsFromVolume(vol);
		compiler.outputFileSystem = memFs as OutputFileSystem;

		compiler.run((err, stats) => {
			if (err || stats?.hasErrors()) {
				reject(err || stats?.toJson().errors);
				return;
			}

			const transformedCode = memFs.readFileSync("/bundle.js", "utf-8");

			resolve(transformedCode.toString());
		});
	});
}

describe("Webpack plugin", () => {
	it("should handle a YAML schema with a reference", async () => {
		const code = await build({
			entry: "./src/fixtures/webpack/main.js",
			plugins: [webpackPlugin()],
		});

		expect(code).toMatchSnapshot();
		expect(code).toContain('type:"object",');
	});

	it("should handle a JSON schema with a reference", async () => {
		const code = await build({
			entry: "./src/fixtures/webpack/main-json.js",
			plugins: [
				webpackPlugin({
					extensions: [".openapi.json"],
				}),
			],
		});

		expect(code).toMatchSnapshot();
		expect(code).toContain('type:"object",');
	});
});
