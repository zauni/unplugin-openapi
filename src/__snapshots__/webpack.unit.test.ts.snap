// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Webpack plugin > should handle a JSON schema with a reference 1`] = `"(()=>{"use strict";console.log({openapi:"3.0.0",info:{title:"My great API",description:"Great description",version:"1.0.0"},paths:{"/my/path":{get:{summary:"Some GET request",responses:{200:{description:"Some response",content:{"application/json":{schema:{type:"object",properties:{someKey:{type:"string"}},required:["someKey"]},example:{someKey:"some value"}}}}}}}}})})();"`;

exports[`Webpack plugin > should handle a YAML schema with a reference 1`] = `"(()=>{"use strict";console.log({openapi:"3.0.0",info:{title:"My great API",description:"Great description",version:"1.0.0"},paths:{"/my/path":{get:{summary:"Some GET request",responses:{200:{description:"Some response",content:{"application/json":{schema:{type:"object",properties:{someKey:{type:"string"}},required:["someKey"]},example:{someKey:"some value"}}}}}}}}})})();"`;
